DEVX-1468: Cloud ETL demo (#429)
confluentinc/examples: cloud-etl/docs/index.rst

.. _cloud-etl:
  
.. toctree::
    :maxdepth: 2

Cloud ETL Demo
==============

> This demo showcases a cloud ETL leveraging all fully-managed services on

Reviewer: maybe consider "a cloud ETL solution"... or maybe people use ETL as a noun but it reads slightly off to me.

Author: 👍 @rspurgeon agree, and fixed

+-----------------------+-----------------------+---------------------+
| GCS/S3/Blob sink      | KSQL tables           | GCS/S3/Blob         |
| connector             | ``COUNT_PER_SOURCE``, |                     |
|                       | ``SUM_PER_SOURCE``    |                     |
+-----------------------+-----------------------+---------------------+

Warning
=======

This demo uses real cloud resources, including that of |ccloud|,
Amazon Kinesis, and one of the cloud storage providers. To avoid
unexpected charges, carefully evaluate the cost of resources before
launching the demo and ensure all resources are destroyed after you are
done running it.

Prerequisites
=============

Cloud services
--------------

-  `Confluent Cloud cluster <https://confluent.cloud>`__: for development only. Do not use a production cluster.
-  `Confluent Cloud KSQL <https://docs.confluent.io/current/quickstart/cloud-quickstart/ksql.html>`__ provisioned in your |ccloud|
-  Download `Confluent Platform <https://www.confluent.io/download/>`__ |release|: for more advanced Confluent CLI functionality (optional)
-  AWS or GCP or Azure access

> Local install

Reviewer: FYC: Maybe the phrase "Local Development Tools" or "Local Tools", I feel "Local Install" is a little vague

Author: 👍 agree, and addressed.

-------------

-  `Confluent Cloud CLI <https://docs.confluent.io/current/quickstart/cloud-quickstart/index.html#step-2-install-the-ccloud-cli>`__
   v0.234.0 or later
-  ``aws`` CLI, properly initialized with your credentials: AWS Kinesis and (optional) if destination is AWS S3

> -  ``gsutils`` CLI, properly initialized with your credentials: (optional) if destination is GPC GCS

Reviewer: ``gsutil`` ? (no 's'). I tested this only because I have it locally but my run thorugh will go to aws

Author: Good catch!!

-  ``az`` CLI, properly initialized with your credentials: (optional) if destination is Azure Blob storage
-  ``jq``
-  ``curl``

Run the Demo
============

Setup
-----

#. By default, the demo reads the configuration parameters for your |ccloud| cluster from a file at ``$HOME/.ccloud/config``. You can change this filename via the parameter ``CONFIG_FILE`` in :devx-examples:`config/demo.cfg|cloud-etl/config/demo.cfg`. Enter the configuration parameters for your |ccloud| cluster:

   .. code:: shell

      $ cat $HOME/.ccloud/config
      bootstrap.servers=<BROKER ENDPOINT>
      ssl.endpoint.identification.algorithm=https
      security.protocol=SASL_SSL
      sasl.mechanism=PLAIN
      sasl.jaas.config=org.apache.kafka.common.security.plain.PlainLoginModule required username\="<API KEY>" password\="<API SECRET>";
      schema.registry.url=https://<SR ENDPOINT>
      schema.registry.basic.auth.user.info=<SR API KEY>:<SR API SECRET>
      basic.auth.credentials.source=USER_INFO
      ksql.endpoint=https://<KSQL ENDPOINT>
      ksql.basic.auth.user.info=<KSQL API KEY>:<KSQL API SECRET>

>    To get the right values for the endpoints in the file above, find them either via the |ccloud| UI or |ccloud| CLI commands. If you have multiple |ccloud| clusters, make sure to use the one with the associated KSQL cluster.

Reviewer: Should we highlight for the user to create and fill in API keys and secrets here as well (or is that coming later somewhere else)?

Author: 👍 excellent point, added.

Warning
=======

This demo uses real cloud resources, including that of |ccloud|,
Amazon Kinesis, and one of the cloud storage providers. To avoid
unexpected charges, carefully evaluate the cost of resources before
launching the demo and ensure all resources are destroyed after you are
done running it.

Prerequisites
=============

Cloud services
--------------

-  `Confluent Cloud cluster <https://confluent.cloud>`__: for development only. Do not use a production cluster.
-  `Confluent Cloud KSQL <https://docs.confluent.io/current/quickstart/cloud-quickstart/ksql.html>`__ provisioned in your |ccloud|
-  Download `Confluent Platform <https://www.confluent.io/download/>`__ |release|: for more advanced Confluent CLI functionality (optional)
-  AWS or GCP or Azure access

Local install
-------------

-  `Confluent Cloud CLI <https://docs.confluent.io/current/quickstart/cloud-quickstart/index.html#step-2-install-the-ccloud-cli>`__
   v0.234.0 or later
-  ``aws`` CLI, properly initialized with your credentials: AWS Kinesis and (optional) if destination is AWS S3
-  ``gsutils`` CLI, properly initialized with your credentials: (optional) if destination is GPC GCS

> -  ``az`` CLI, properly initialized with your credentials: (optional) if destination is Azure Blob

Reviewer 2: ```suggestion
-  ``az`` CLI, properly initialized with your credentials: (optional) if destination is Azure Blob storage
```

-  ``jq``
-  ``curl``

Run the Demo
============

Setup
-----

> #. Initialize a properties file at ``$HOME/.ccloud/config`` with the

Reviewer 2: maybe: Modify the ``$HOME/.ccloud/config`` properties file. Enter the configuration parameters for your |ccloud| cluster.

Author: Ack, modified

   following configuration parameters for your |ccloud| cluster:

   .. code:: shell

      $ cat $HOME/.ccloud/config
      bootstrap.servers=<BROKER ENDPOINT>
      ssl.endpoint.identification.algorithm=https
      security.protocol=SASL_SSL
      sasl.mechanism=PLAIN
      sasl.jaas.config=org.apache.kafka.common.security.plain.PlainLoginModule required username\="<API KEY>" password\="<API SECRET>";
      schema.registry.url=https://<SR ENDPOINT>
      schema.registry.basic.auth.user.info=<SR API KEY>:<SR API SECRET>
      basic.auth.credentials.source=USER_INFO
      ksql.endpoint=https://<KSQL ENDPOINT>
      ksql.basic.auth.user.info=<KSQL API KEY>:<KSQL API SECRET>

>    To get the right values for the endpoints in the file above, find them either via the |ccloud| UI or |ccloud| CLI commands. If you have multiple |ccloud| clusters, make sure to use the one with the associated KSQL cluster.

Reviewer 2: ```suggestion
   To get the right values for the endpoints in the file above, find them either via the |ccloud| UI or |ccloud| CLI commands as shown below. If you have multiple |ccloud| clusters, make sure to use the one with the associated KSQL cluster.
```

   ::

   # Login
   ccloud login --url https://confluent.cloud

   # BROKER ENDPOINT
   ccloud kafka cluster list
   ccloud kafka cluster use
   ccloud kafka cluster describe

   # SR ENDPOINT
   ccloud schema-registry cluster describe

   # KSQL ENDPOINT
   ccloud ksql app list

#. Clone the `examples GitHub repository <https://github.com/confluentinc/examples>`__ and check out the :litwithvars:`|release|-post` branch.

   .. codewithvars:: bash

     git clone https://github.com/confluentinc/examples
     cd examples
     git checkout |release|-post

#. Change directory to the Cloud ETL demo at ``cloud-etl``:

   .. codewithvars:: bash

     cd cloud-etl

#. Modify the demo configuration file at :devx-examples:`config/demo.cfg|cloud-etl/config/demo.cfg` for your particular cloud storage provider and other demo parameters.

#. Log into |ccloud| with the command ``ccloud login``, and use
   your |ccloud| username and password.

   ::

>    ccloud login --url https://confluent.cloud

Reviewer 2: is the --url required here?

Reviewer 2: does the user have to configure their ccloud environment after logging in?

   192.168.1.2 {"EVENTSOURCEIP":{"string":"192.168.1.2"},"KSQL_COL_1":{"long":19}}
   192.168.1.2 {"EVENTSOURCEIP":{"string":"192.168.1.2"},"KSQL_COL_1":{"long":22}}

   Objects in Cloud storage az:

   topics/COUNT_PER_SOURCE/year=2020/month=02/day=11/hour=18/COUNT_PER_SOURCE+1+0000000000.bin
   topics/COUNT_PER_SOURCE/year=2020/month=02/day=11/hour=18/COUNT_PER_SOURCE+1+0000001000.bin
   topics/COUNT_PER_SOURCE/year=2020/month=02/day=11/hour=18/COUNT_PER_SOURCE+1+0000002000.bin
   topics/COUNT_PER_SOURCE/year=2020/month=02/day=11/hour=18/COUNT_PER_SOURCE+1+0000003000.bin
   topics/COUNT_PER_SOURCE/year=2020/month=02/day=11/hour=18/COUNT_PER_SOURCE+1+0000004000.bin
   topics/COUNT_PER_SOURCE/year=2020/month=02/day=11/hour=18/COUNT_PER_SOURCE+3+0000000000.bin
   topics/COUNT_PER_SOURCE/year=2020/month=02/day=11/hour=18/COUNT_PER_SOURCE+3+0000001000.bin
   topics/COUNT_PER_SOURCE/year=2020/month=02/day=11/hour=18/COUNT_PER_SOURCE+3+0000002000.bin
   topics/COUNT_PER_SOURCE/year=2020/month=02/day=11/hour=18/COUNT_PER_SOURCE+3+0000003000.bin
   topics/COUNT_PER_SOURCE/year=2020/month=02/day=11/hour=18/COUNT_PER_SOURCE+3+0000004000.bin

Stop
----

> #. Stop the demo and clean up all the resources, delete Kafka topics, delete the fully-managed connectors, delete the data in the cloud storage:

Reviewer 2: ```suggestion
#. Stop the demo and clean up all the resources, delete Kafka topics, delete the fully-managed connectors, and delete the provider cloud storage resources.
```
